Feature: allow the use of arbitrary brackets (e.g. tags) via aliases

Daxtorim opened the pull request:
This PR builds on top of #37 (the first four commits). I just split this into 2 PRs since they are each already pretty large on their own.

This PR starts out with a refactor of parser.lua by consolidating all the `get_[non_]nested_pair()` and `get_surrounding_function()` functions and gave them the ability to look for multi-character brackets. I also split out another function with code that was copy-pasted SIX (!) times in there.

While at it, I also added the ability to remove and thus change the surrounding spacing. In normal mode, `sr()` will now remove spaces after the opening bracket and before the closing bracket and then add the brackets back without the space.
(was mentioned in #31 [here](https://github.com/blackCauldron7/surround.nvim/issues/31#issuecomment-1001875999))

Diff of just the refactor and space changing fix: https://github.com/blackCauldron7/surround.nvim/compare/96b5c71b2b680b000728b89faa83a7f8467a9bf6...cfd6f16a367c7f55522238379a1eefcaec2cba19

---

After that everything is to allow for arbitrary brackets via aliases. The new way to setup the `pairs` table would be:
```lua
pairs = {
  nestable = {
    b = { "(", ")" },
    s = { "[", "]" },
    B = { "{", "}" },
    a = { "<", ">" }
    },
  linear = {
    q = { "'", "'" },
    t = { "`", "`" },
    d = { '"', '"' }
  },
}
```
These are also the new default values. If the pair is composed of only single-character brackets, either bracket can be used to add/replace/delete surroundings. If the brackets are build out of multiple characters, the key of the table must be used as an alias (only single-character keys allowed for now)
Example:
```lua
pairs = {
  nestable = {
    d = { "<div>", "</div>" },
  },
}
```
This setting would allow you to add surround div tags by using `d` as the alias (i.e. `saiwd` will surround the word under the cursor with this tag). There is also a new setting `space_on_alias`, that should be self-explanatory here.
Closes #36 
Partially implements #31 (no dynamic "buns" 😢, I have no idea right now how to approach that to make it work with `replace()` and `delete()`)

Diff: https://github.com/blackCauldron7/surround.nvim/compare/a6ba7a151dd316069d1e1252eabbeff637bd0744...951e44f3890b44d2233e7988b6e2fa29e2b15757

blackCauldron7 commented:
Hey! Did you robustly test this? I can then probably just merge this. Did a simple code review everything seems good at first look atleast

Daxtorim commented:
I rebased the branch on master again and added a new commit that fixes behavior when the cursor is already inside one of the brackets.

I did test add/replace/delete functionality with various different pairs and they all seem to work.

---

The one thing that changed its behavior is the search for non-nestable pairs. On master currently it will only search the line where the cursor is on for non-nestable pairs. That implementation has the nice property of only allowing the deletion of actually connected pairs of quotes. The downside is that it only looks on the one line the cursor was on.

By basically using the same implementation for nestable and non-nestable pairs here I inadvertently switched those two properties. You can now delete quotes that are on different lines, but there is no guarantee that those two quotes are actually paired.

Pressing `sdd` to delete the double quotes when the cursor is over the `m` in `some` will result in one long string:
```lua
--                             |cursor
var = "Start of string; " .. some_variable .. "; end of string"
-- Result after pressing `sdd`:
var = "Start of string;  .. some_variable .. ; end of string"
```
So do you think that safety is more desirable over the ability to change quotes of multi-line strings (like python docstrings)?
We can't have both as that would require searching through the entire *file*.

blackCauldron7 commented:
I would personally go for the safety as broken code is worse than not being able to delete multiline linear pairs.

Maybe a poll on reddit or twitter to decide XD?

Daxtorim commented:
I don't know how much that "safety" is really worth. How difficult can it be to not delete quotes with the cursor in random places? Since this is neovim only and pretty programming oriented, anyone using this has very likely syntax highlighting active or even LSP servers showing them error messages because of the missing quotes. *You can still undo your error with one button press.*

I kinda knew that I would be removing the "safety" feature at the beginning and thought that having multi-line support would be better since the safety is not bullet proof and the multi-line support would allow for some other applications than just quotes around strings.

But to be honest, now I can't actually come up with an example for other non-nestable pairs that aren't just python docstrings or triple backticks for code blocks. And how often do you mess with the quotes around either of those?
Maybe someone else has actual use-cases for multi-line support. <sup>(Maybe future features?)</sup>

---

>Maybe a poll on reddit or twitter to decide XD?

I don't know how much reach you have on either, but sure, try your luck :rofl:.

blackCauldron7 commented:
can you separate the quotes into another PR?

Daxtorim commented:
Not really. This is a side effect of cf55915. Undoing that would lead to massive conflicts as the other commits build on that.

I could try to separate the functions again a little, but that would be a new commit that the new PR would just undo again.

Daxtorim commented:
I restored the old behavior again.

blackCauldron7 commented:
Thanks for the help :smile:
